Bottom: 4b6adce26ba8f646a5c258134794671aa8803aba
Top:    94ac9194d21b2a3a8ea6b3ab15468b3c8632302a
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2008-09-30 22:52:44 +0100

Don't bother unsubtracting complex polygons when doing incremental updates

Since the polygon is re-cleared against any objects existing within the
bounding box of the update region, there is no point wasting CPU cycles
intersecting a more complex polygon than necessary. Just take the UNION
of the polygon being unsubtracted, and the bounding box of the hole to
be filed in.




---

diff --git a/src/misc.c b/src/misc.c
index 2153a1a..df34363 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -465,6 +465,14 @@ SetTextBoundingBox (FontTypePtr FontPtr, TextTypePtr Text)
       RotateBoxLowLevel (&Text->BoundingBox,
                          Text->X, Text->Y, Text->Direction);
     }
+
+  /* the bounding box covers the extent of influence
+   * so it must include the clearance values too
+   */
+  Text->BoundingBox.X1 -= PCB->Bloat;
+  Text->BoundingBox.Y1 -= PCB->Bloat;
+  Text->BoundingBox.X2 += PCB->Bloat;
+  Text->BoundingBox.Y2 += PCB->Bloat;
 }
 
 /* ---------------------------------------------------------------------------
diff --git a/src/polygon.c b/src/polygon.c
index 50912b2..d67b918 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -65,6 +65,8 @@ RCSID ("$Id$");
 
 #define ROUND(x) ((long)(((x) >= 0 ? (x) + 0.5  : (x) - 0.5)))
 
+#define UNSUBTRACT_BLOAT 10
+
 /* ---------------------------------------------------------------------------
  * local prototypes
  */
@@ -590,6 +592,13 @@ PinPoly (PinType * pin, BDimension thick, BDimension clear)
   return CirclePoly (pin->X, pin->Y, size);
 }
 
+POLYAREA *
+BoxPolyBloated (BoxType *box, BDimension bloat)
+{
+  return RectPoly (box->X1 - bloat, box->X2 + bloat,
+                   box->Y1 - bloat, box->Y2 + bloat);
+}
+
 /* remove the pin clearance from the polygon */
 static int
 SubtractPin (DataType * d, PinType * pin, LayerType * l, PolygonType * p)
@@ -854,46 +863,51 @@ UnsubtractPin (PinType * pin, LayerType * l, PolygonType * p)
   POLYAREA *np;
 
   /* overlap a bit to prevent gaps from rounding errors */
-  np = PinPoly (pin, pin->Thickness, pin->Clearance * 1.1);
+  np = BoxPolyBloated (&pin->BoundingBox, UNSUBTRACT_BLOAT);
 
   if (!np)
     return 0;
   if (!Unsubtract (np, p))
     return 0;
-  clearPoly (PCB->Data, l, p, (const BoxType *) pin,
-             0.1 * (pin->Thickness + pin->Clearance));
+  clearPoly (PCB->Data, l, p, (const BoxType *) pin, 2 * UNSUBTRACT_BLOAT);
   return 1;
 }
 
 static int
 UnsubtractArc (ArcType * arc, LayerType * l, PolygonType * p)
 {
-  POLYAREA *np = NULL;
+  POLYAREA *np;
 
   if (!TEST_FLAG (CLEARLINEFLAG, arc))
     return 0;
-  /* overlap a bit to prevent notches from rounding errors */
-  if (!(np = ArcPoly (arc, arc->Thickness + arc->Clearance + 100)))
+
+  /* overlap a bit to prevent gaps from rounding errors */
+  np = BoxPolyBloated (&arc->BoundingBox, UNSUBTRACT_BLOAT);
+
+  if (!np)
     return 0;
   if (!Unsubtract (np, p))
     return 0;
-  clearPoly (PCB->Data, l, p, (const BoxType *) arc, 50);
+  clearPoly (PCB->Data, l, p, (const BoxType *) arc, 2 * UNSUBTRACT_BLOAT);
   return 1;
 }
 
 static int
 UnsubtractLine (LineType * line, LayerType * l, PolygonType * p)
 {
-  POLYAREA *np = NULL;
+  POLYAREA *np;
 
   if (!TEST_FLAG (CLEARLINEFLAG, line))
     return 0;
+
   /* overlap a bit to prevent notches from rounding errors */
-  if (!(np = LinePoly (line, line->Thickness + line->Clearance + 100)))
+  np = BoxPolyBloated (&line->BoundingBox, UNSUBTRACT_BLOAT);
+
+  if (!np)
     return 0;
   if (!Unsubtract (np, p))
     return 0;
-  clearPoly (PCB->Data, l, p, (const BoxType *) line, 50);
+  clearPoly (PCB->Data, l, p, (const BoxType *) line, 2 * UNSUBTRACT_BLOAT);
   return 1;
 }
 
@@ -901,40 +915,34 @@ static int
 UnsubtractText (TextType * text, LayerType * l, PolygonType * p)
 {
   POLYAREA *np;
-  const BoxType *b = &text->BoundingBox;
 
   if (!TEST_FLAG (CLEARLINEFLAG, text))
     return 0;
-  if (!(np = RoundRect (b->X1, b->X2, b->Y1, b->Y2, PCB->Bloat + 100)))
+
+  /* overlap a bit to prevent notches from rounding errors */
+  np = BoxPolyBloated (&text->BoundingBox, UNSUBTRACT_BLOAT);
+
+  if (!np)
     return -1;
   if (!Unsubtract (np, p))
     return 0;
-  clearPoly (PCB->Data, l, p, (const BoxType *) text, 101);
+  clearPoly (PCB->Data, l, p, (const BoxType *) text, 2 * UNSUBTRACT_BLOAT);
   return 1;
 }
 
 static int
 UnsubtractPad (PadType * pad, LayerType * l, PolygonType * p)
 {
-  POLYAREA *np = NULL;
+  POLYAREA *np;
+
+  /* overlap a bit to prevent notches from rounding errors */
+  np = BoxPolyBloated (&pad->BoundingBox, UNSUBTRACT_BLOAT);
 
-  if (TEST_FLAG (SQUAREFLAG, pad))
-    {
-      if (!
-          (np = SquarePadPoly (pad, pad->Thickness + pad->Clearance + 100)))
-        return 0;
-    }
-  else
-    {
-      if (!
-          (np = LinePoly ((LineType *) pad, pad->Thickness + pad->Clearance + 100)))
-        return 0;
-    }
   if (!np)
     return 0;
   if (!Unsubtract (np, p))
     return 0;
-  clearPoly (PCB->Data, l, p, (const BoxType *) pad, 100);
+  clearPoly (PCB->Data, l, p, (const BoxType *) pad, 2 * UNSUBTRACT_BLOAT);
   return 1;
 }
 
diff --git a/src/polygon.h b/src/polygon.h
index a7b9da7..721ba95 100644
--- a/src/polygon.h
+++ b/src/polygon.h
@@ -50,6 +50,7 @@ POLYAREA * OctagonPoly(LocationType x, LocationType y, BDimension radius);
 POLYAREA * LinePoly(LineType *l, BDimension thick);
 POLYAREA * ArcPoly(ArcType *l, BDimension thick);
 POLYAREA * PinPoly(PinType *l, BDimension thick, BDimension clear);
+POLYAREA * BoxPolyBloated (BoxType *box, BDimension radius);
 void frac_circle (PLINE *, LocationType, LocationType, Vector, int);
 int InitClip(DataType *d, LayerType *l, PolygonType *p);
 void RestoreToPolygon(DataType *, int, void *, void *);
